snowplow-tracker/src/main/java/com/snowplowanalytics/snowplow/internal/session/Session.java

```diff
@@ -292,47 +312,75 @@ int getForegroundIndex() {
-        return sessionValues;
-    }
-
     /**
      * Gets the session information from a file.
      *
      * @return a map or null.
      */
     @Nullable
-    private Map<String, Object> getSessionFromFile(@NonNull Context context) {
-        return FileStore.getMapFromFile(
+    private Map<String, Object> getSessionMapFromLegacyTrackerV1(@NonNull Context context) {
+        Map<String, Object> sessionMap = FileStore.getMapFromFile(
                 TrackerConstants.SNOWPLOW_SESSION_VARS,
                 context);
+        sessionMap.put(Parameters.SESSION_FIRST_ID, "");
+        sessionMap.put(Parameters.SESSION_PREVIOUS_ID, null);
+        sessionMap.put(Parameters.SESSION_STORAGE, "LOCAL_STORAGE");
+        return sessionMap;
     }
 
     @Nullable
-    private SharedPreferences getSessionFromSharedPreferences(@NonNull Context context, @NonNull String sessionVarsName) {
+    private Map<String, Object> getSessionMapFromLegacyTrackerV2(@NonNull Context context, @NonNull String sessionVarsName) {
         StrictMode.ThreadPolicy oldPolicy = StrictMode.allowThreadDiskReads();
         try {
             SharedPreferences sharedPreferences = context.getSharedPreferences(sessionVarsName, Context.MODE_PRIVATE);
-            if (sharedPreferences.contains(Parameters.SESSION_ID)) {
-                return sharedPreferences;
-            } else {
+            if (!sharedPreferences.contains(Parameters.SESSION_ID)) {
                 sharedPreferences = context.getSharedPreferences(TrackerConstants.SNOWPLOW_SESSION_VARS, Context.MODE_PRIVATE);
-                if (sharedPreferences.contains(Parameters.SESSION_ID)) {
-                    return sharedPreferences;
+                if (!sharedPreferences.contains(Parameters.SESSION_ID)) {
+                    return null;
                 }
             }
+            // Create map used as initial session state
+            Map<String, Object> sessionMap = new HashMap<>();
+            String sessionId = sharedPreferences.getString(Parameters.SESSION_ID, null);
+            if (sessionId == null) return null;
+            sessionMap.put(Parameters.SESSION_ID, sessionId);
+
+            String userId = sharedPreferences.getString(Parameters.SESSION_USER_ID, null);
+            if (userId == null) return null;
+            sessionMap.put(Parameters.SESSION_USER_ID, userId);
+
+            int sessionIndex = sharedPreferences.getInt(Parameters.SESSION_INDEX, 0);
+            sessionMap.put(Parameters.SESSION_INDEX, sessionIndex);
+
+            sessionMap.put(Parameters.SESSION_FIRST_ID, "");
+            sessionMap.put(Parameters.SESSION_PREVIOUS_ID, null);
+            sessionMap.put(Parameters.SESSION_STORAGE, "LOCAL_STORAGE");
+            return sessionMap;
+        } finally {
+            StrictMode.setThreadPolicy(oldPolicy);
+        }
+    }
+
+    @Nullable
+    private Map<String, Object> getSessionMapFromLegacyTrackerV3(@NonNull Context context, @NonNull String sessionVarsName) {
```

Comment:
This is quite off-topic but just wanted to suggest that, in the future, we can look into deserialization of maps or JSONs into POJO (maybe using the Jackson library). Probably it's just about preference, but I find this dictionary parsing a bit hard to follow and check. I think defining a simple class for what is expected and then mapping the Map to it would be more explicit. I'm always happier to work with POJO than maps.

Reply:
I agree with the general point and some of the previous refactoring went exactly in this direction.
I have just a couple of concerns: 
- I wouldn't add other dependencies in the library. Jackson is definitely cool and widely used but we don't own the final app where the tracker will be instrumented and any dependency we add can be a source of issues in the app and for us with the maintenance.
- SessionState could be considered the POJO but I didn't want to add the serialization code there. On iOS tracker I created SPDataPersitence and that seems a bit cleaner than this.

My doubt is, If we avoid Jackson lib, would the code we write for serialization/deserialization be much different than this?
I think much of this complexity would fade away when (if) we transform the Session manager in a FSM like I did for ScreenContext and DeepLinks. Unfortunately it was off-topic and a too long shot for this task.

Reply:
We should discuss this further somewhere when we have a chance. I agree that introducing additional dependencies is not ideal. Many modern languages (Swift, Kotlin) already have support for this built in. But if there is no easy way currently to do it in Java, it might be better to hold off on the idea rather than adding a dependency.

I do believe that there is a difference between writing code to parse a dictionary (or JSON) and mapping a dictionary onto an interface (POJO class). If we avoid writing the parsing code and only define what we expect and let the framework map it, it can reduce the potential for bugs, make the code more explicit and easier to read. But I definitely agree that it is not the only way to write good code, we sure can write easy to follow code with the current approach as well.